File: backend/app/schemas/attribute.py
from datetime import datetime

from pydantic import BaseModel, field_validator

from backend.app.models.attribute import DataType, OwnerType


class AttributeConstraints(BaseModel):
    required: bool = False
    default: str | int | float | bool | None = None
    enum_values: list[str] | None = None
    validation_regex: str | None = None
    min_value: float | None = None
    max_value: float | None = None
    min_length: int | None = None
    max_length: int | None = None


class AttributeDefinitionCreate(BaseModel):
    name: str
    display_name: str
    data_type: DataType
    owner_type: OwnerType
    campaign_id: int | None = None
    constraints: AttributeConstraints | None = None
    is_fixed: bool = False


class AttributeDefinitionUpdate(BaseModel):
    display_name: str | None = None
    constraints: AttributeConstraints | None = None


class AttributeDefinitionRead(BaseModel):
    id: int
    name: str
    display_name: str
    data_type: DataType
    owner_type: OwnerType
    campaign_id: int | None = None
    constraints: AttributeConstraints | None = None
    is_fixed: bool
    created_at: datetime

    model_config = {"from_attributes": True}

    @field_validator("constraints", mode="before")
    @classmethod
    def parse_constraints(cls, v):
        if isinstance(v, dict):
            return AttributeConstraints(**v)
        return v

File: frontend/src/types/attribute.ts
export type DataType =
  | 'string' | 'int' | 'float' | 'bool'
  | 'date' | 'datetime' | 'enum' | 'list'
  | 'phone' | 'email' | 'url'

export type OwnerType = 'contact' | 'campaign_member'

export interface AttributeConstraints {
  required?: boolean
  default?: string | number | boolean | null
  enum_values?: string[] | null
  validation_regex?: string | null
  min_value?: number | null
  max_value?: number | null
  min_length?: number | null
  max_length?: number | null
}

export interface AttributeDefinition {
  id: number
  name: string
  display_name: string
  data_type: DataType
  owner_type: OwnerType
  campaign_id: number | null
  constraints: AttributeConstraints | null
  is_fixed: boolean
  created_at: string
}

export interface AttributeDefinitionCreate {
  name: string
  display_name: string
  data_type: DataType
  owner_type: OwnerType
  campaign_id?: number | null
  constraints?: AttributeConstraints | null
  is_fixed?: boolean
}

export interface AttributeDefinitionUpdate {
  display_name?: string
  constraints?: AttributeConstraints | null
}
